packages/astro/src/core/cookies/cookies.ts

```diff
@@ -0,0 +1,236 @@
+			return isSetValue;
+		}
+		const values = this.#ensureParsed();
+		return !!values[key];
+	}
+
+	/**
+	 * Astro.cookies.set(key, value) is used to set a cookie's value. If provided
+	 * an object it will be stringified via JSON.stringify(value). Additionally you
+	 * can provide options customizing how this cookie will be set, such as setting httpOnly
+	 * in order to prevent the cookie from being read in client-side JavaScript.
+	 * @param key The name of the cookie to set.
+	 * @param value A value, either a string or other primitive or an object.
+	 * @param options Options for the cookie, such as the path and security settings.
+	 */
+	set(key: string, value: string | Record<string, any>, options?: AstroCookieSetOptions): void {
+		let serializedValue: string;
+		if(typeof value === 'string') {
+			serializedValue = value;
+		} else {
+			// Support stringifying JSON objects for convenience. First check that this is
+			// a plain object and if it is, stringify. If not, allow support for toString() overrides.
+			let toStringValue = value.toString();
+			if(toStringValue === Object.prototype.toString.call(value)) {
+				serializedValue = JSON.stringify(value);
+			} else {
+				serializedValue = toStringValue;
+			}
+		}
+
+		let expires: Date | undefined = undefined;
+		if(options?.expires) {
+			let rawExpires = options.expires;
+			switch(typeof rawExpires) {
+				case 'string': {
+					let numberOfMs = ms(rawExpires);
+					if(numberOfMs === undefined) {
+						if(rawExpires.includes('month')) {
+							throw new Error(`Cannot convert months because there is no fixed duration. Use days instead.`);
+						} else {
+							throw new Error(`Unable to convert expires expression [${rawExpires}]`);
+						}
+					}
+					let now = Date.now();
+					expires = new Date(now + numberOfMs);
+					break;
+				}
+				case 'number': {
+					expires = new Date(rawExpires);
+					break;
+				}
+				default: {
+					expires = rawExpires;
+					break;
+				}
+			}
+		}
+
+		const serializeOptions: CookieSerializeOptions = {};
+		if(options) {
+			Object.assign(serializeOptions, options, {
+				expires
+			});
+		}
+
+		this.#ensureOutgoingMap().set(key, [
+			serializedValue,
+			serialize(key, serializedValue, serializeOptions),
+			true
+		]);
+	}
+
+	/**
+	 * Astro.cookies.header() returns an iterator for the cookies that have previously
+	 * been set by either Astro.cookies.set() or Astro.cookies.delete().
+	 * This method is primarily used by adapters to set the header on outgoing responses.
+	 * @returns 
+	 */
+	*headers(): Generator<string, void, unknown> {
+		if(this.#outgoing == null) return;
+		for(const [,value] of this.#outgoing) {
+			yield value[1];
+		}
+	}
+
+	#ensureParsed(): Record<string, string> {
+		if(!this.#requestValues) {
+			this.#parse();
+		}
+		if(!this.#requestValues) {
+			this.#requestValues = {};
+		}
+		return this.#requestValues;
+	}
+
+	#ensureOutgoingMap(): Map<string, [string, string, boolean]> {
```

Comment:
Is there a reason to not initialize `#outgoing` in the constructor? It doesn’t seem to ever get re-set to `null`. That would avoid the need for this method.

Reply:
I'm trying to do everything as lazy as possible in this class to avoid unnecessary work. The majority of pages don't read cookies, so we should lazily initiate it. The majority of pages that do read cookies don't also set them (mostly POST requests do that), so we should wait until we know that one needs to be set to construct these objects.